src/auto-trace.js

```diff
@@ -0,0 +1,132 @@
+const globalMiddlewares = [];
+
+/**
+ * Adds a middleware function that will be called on all errors before being handled by auto-trace
+ *
+ * @param {function (Object err) => Object newErr } middlewareFn 
+ */
+export function addGlobalMiddleware(middlewareFn) {
+	globalMiddlewares.push(middlewareFn)
+}
+
+/**
+ * Wraps rawError in an Error object (if typeOf rawError =! Error) using the async stacktrace
+ * Calls globalMiddleware and extraMiddleware functions on rawError before wrapping in Error object
+ * In most cases the async stacktrace is most useful - it contains the stacktrace before switching context to caller timeout
+ *
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError before being handled
+ * @returns {Error} 
+ */
+export function asyncStacktrace(callback = ()=>{}, ...extraMiddlewares) {
+	const stacktraceErr = new Error();
+
+	return (rawError) => {
+		const err = rawError ? rawError : new Error();
+		const middlewareErr = executeMiddleware(err, extraMiddlewares);
+		callback(wrapObjectWithError(middlewareErr, stacktraceErr));
+	};	
+}
+
+/**
+ * Wraps rawError in an Error object (if typeOf rawError =! Error) using the async stacktrace
+ * Calls globalMiddleware and extraMiddleware functions on rawError before wrapping in Error object 
+ * Behaves exactly as asyncStacktrace, but throws the error object instead of returning.
+ * In most cases the async stacktrace is most useful - it contains the stacktrace before switching context to caller timeout 
+ *
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError before being handled
+ * @throws {Error} 
+ */
+export function throwAsyncStacktrace(...extraMiddlewares) {
+	const stacktraceErr = new Error();
+
+	return (rawError) => {
+		const err = rawError ? rawError : new Error();
```

Comment:
this isn't something that needs to be changed, but have you ever tried out `const err = rawError || new Error()`. It is even shorter than the ternary and is basically an if statement.